core/acl/aclservice.go

```diff
@@ -0,0 +1,477 @@
+	if err != nil {
+		return err
+	}
+	acl := acceptSpace.CommonSpace().Acl()
+	acl.RLock()
+	recs, err := acl.AclState().JoinRecords(false)
+	if err != nil {
+		acl.RUnlock()
+		return err
+	}
+	// TODO: change this logic to use RequestJoin objects
+	var recId string
+	for _, rec := range recs {
+		if rec.RequestIdentity.Equals(identity) {
+			recId = rec.RecordId
+			break
+		}
+	}
+	acl.RUnlock()
+	if recId == "" {
+		return fmt.Errorf("%w with identity: %s", ErrNoSuchUser, identity.Account())
+	}
+	cl := acceptSpace.CommonSpace().AclClient()
+	var aclPerms list.AclPermissions
+	switch permissions {
+	case model.ParticipantPermissions_Reader:
+		aclPerms = list.AclPermissionsReader
+	case model.ParticipantPermissions_Writer:
+		aclPerms = list.AclPermissionsWriter
+	}
+	err = cl.AcceptRequest(ctx, list.RequestAcceptPayload{
+		RequestRecordId: recId,
+		Permissions:     aclPerms,
+	})
+	if err != nil {
+		return fmt.Errorf("%w, %w", ErrAclRequestFailed, err)
+	}
+	return nil
+}
+
+type InviteInfo struct {
+	InviteFileCid string
+	InviteFileKey string
+}
+
+func (a *aclService) buildInvite(ctx context.Context, space clientspace.Space, inviteKey crypto.PrivKey) (*model.Invite, error) {
+	invitePayload, err := a.buildInvitePayload(ctx, space, inviteKey)
+	if err != nil {
+		return nil, fmt.Errorf("build invite payload: %w", err)
+	}
+	invitePayloadRaw, err := proto.Marshal(invitePayload)
+	if err != nil {
+		return nil, fmt.Errorf("marshal invite payload: %w", err)
+	}
+	invitePayloadSignature, err := a.accountService.SignData(invitePayloadRaw)
+	if err != nil {
+		return nil, fmt.Errorf("sign invite payload: %w", err)
+	}
+	return &model.Invite{
+		Payload:   invitePayloadRaw,
+		Signature: invitePayloadSignature,
+	}, nil
+}
+
+func (a *aclService) buildInvitePayload(ctx context.Context, space clientspace.Space, inviteKey crypto.PrivKey) (*model.InvitePayload, error) {
+	profile, err := a.accountService.ProfileInfo()
+	if err != nil {
+		return nil, fmt.Errorf("get profile info: %w", err)
+	}
+	rawInviteKey, err := inviteKey.Marshall()
+	if err != nil {
+		return nil, fmt.Errorf("marshal invite priv key: %w", err)
+	}
+	invitePayload := &model.InvitePayload{
+		SpaceId:         space.Id(),
+		CreatorIdentity: a.accountService.AccountID(),
+		CreatorName:     profile.Name,
+		InviteKey:       rawInviteKey,
+	}
+	err = space.Do(space.DerivedIDs().Workspace, func(sb smartblock.SmartBlock) error {
+		details := sb.Details()
+		invitePayload.SpaceName = pbtypes.GetString(details, bundle.RelationKeyName.String())
+		iconObjectId := pbtypes.GetString(details, bundle.RelationKeyIconImage.String())
+		if iconObjectId != "" {
+			iconCid, iconEncryptionKeys, err := a.fileAcl.GetInfoForFileSharing(ctx, iconObjectId)
+			if err != nil {
+				return fmt.Errorf("get icon info: %w", err)
+			}
+			invitePayload.SpaceIconCid = iconCid
+			invitePayload.SpaceIconEncryptionKeys = iconEncryptionKeys
+		}
+		return nil
+	})
+	if err != nil {
```

Comment:
what if icon broken and not possible to load? maybe we can create the invite without it in this case (just an idea, need confirmation)

Reply:
fixed, thanks